File: src/features/library/review.ts
/* MISSION-074 — Per-media review & notes. Reads the review row for the detail
   page's Review tab, saves it (validating the server-side invariants), clears
   it, and manages the media's personal tags. MISSION-079 adds the
   mood/pace/content-warning metadata and the content-warning acknowledgment.
   The save/acknowledge responses are seeded into the review cache so the tab
   and the detail-page badges reflect the write immediately. */

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import {
  media_add_tag,
  media_remove_tag,
  media_tags,
  review_acknowledge_warnings,
  review_delete,
  review_get,
  review_list,
  review_save,
  type MediaTagView,
} from "@/api";
import { queryKeys } from "@/api";

/** The full review payload the Review tab submits. */
export interface SaveReviewInput {
  media_id: string;
  rating: number | null;
  review: string | null;
  short_review: string | null;
  notes: string | null;
  favorite: boolean;
  is_spoiler: boolean;
  /** Canonical mood keys (MISSION-079). */
  moods: string[];
  pace: string | null;
  /** Canonical content-warning keys (MISSION-079). */
  content_warnings: string[];
}

/** Read the review row for one media (`null` when unreviewed). `enabled` lets
    callers defer the fetch until a media id is available (MISSION-079). */
export function useReviewQuery(mediaId: string, enabled = true) {
  return useQuery({
    queryKey: queryKeys.review.forMedia(mediaId),
    queryFn: () => review_get({ mediaId: mediaId }),
    enabled,
  });
}

/** Every review in the library with its media's display fields, newest first
    (MISSION-144) — feeds the aggregate Reviews hub. */
export function useReviewsQuery() {
  return useQuery({
    queryKey: queryKeys.review.list(),
    queryFn: () => review_list(),
  });
}

/** Save (create or update) a media's review; seeds the review cache. */
export function useSaveReview() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: SaveReviewInput) =>
      review_save({
        mediaId: input.media_id,
        rating: input.rating,
        review: input.review,
        shortReview: input.short_review,
        notes: input.notes,
        favorite: input.favorite,
        isSpoiler: input.is_spoiler,
        moods: input.moods,
        pace: input.pace,
        contentWarnings: input.content_warnings,
      }),
    onSuccess: (view) => {
      queryClient.setQueryData(queryKeys.review.forMedia(view.media_id), view);
      void queryClient.invalidateQueries({ queryKey: queryKeys.review.list() });
      void queryClient.invalidateQueries({ queryKey: queryKeys.media.details() });
      void queryClient.invalidateQueries({ queryKey: queryKeys.dashboard.all() });
    },
  });
}

/** Acknowledge a media's current content-warning set; seeds the cache. */
export function useAcknowledgeWarnings() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (mediaId: string) => review_acknowledge_warnings({ mediaId: mediaId }),
    onSuccess: (view) => {
      queryClient.setQueryData(queryKeys.review.forMedia(view.media_id), view);
      void queryClient.invalidateQueries({ queryKey: queryKeys.review.list() });
    },
  });
}

/** Delete a media's review; seeds an empty cache entry. */
export function useDeleteReview() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (mediaId: string) => review_delete({ mediaId: mediaId }),
    onSuccess: (_void, mediaId) => {
      queryClient.setQueryData(queryKeys.review.forMedia(mediaId), null);
      void queryClient.invalidateQueries({ queryKey: queryKeys.review.list() });
      void queryClient.invalidateQueries({ queryKey: queryKeys.media.details() });
      void queryClient.invalidateQueries({ queryKey: queryKeys.dashboard.all() });
    },
  });
}

/** The personal tags linked to one media (MISSION-074). */
export function useMediaTagsQuery(mediaId: string) {
  return useQuery({
    queryKey: queryKeys.media.tags(mediaId),
    queryFn: () => media_tags({ mediaId: mediaId }),
  });
}

function seedTags(
  queryClient: ReturnType<typeof useQueryClient>,
  mediaId: string,
  tags: MediaTagView[],
) {
  queryClient.setQueryData(queryKeys.media.tags(mediaId), tags);
  void queryClient.invalidateQueries({ queryKey: queryKeys.media.details() });
}

/** Add a personal tag to one media; seeds the updated tag list. */
export function useAddMediaTag() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ mediaId, tag }: { mediaId: string; tag: string }) =>
      media_add_tag({ mediaId: mediaId, tag }),
    onSuccess: (tags, { mediaId }) => seedTags(queryClient, mediaId, tags),
  });
}

/** Remove a personal tag from one media; seeds the updated tag list. */
export function useRemoveMediaTag() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ mediaId, tagId }: { mediaId: string; tagId: string }) =>
      media_remove_tag({ mediaId: mediaId, tagId: tagId }),
    onSuccess: (tags, { mediaId }) => seedTags(queryClient, mediaId, tags),
  });
}

File: src-tauri/src/commands/review.rs
//! Review commands (MISSION-074). Thin handlers over
//! `application::review_service` — read/save/clear a media's user-owned review
//! and acknowledge its content-warning set (MISSION-079).

use sqlx::SqlitePool;
use tauri::command;
use tauri::State;
use tracing::info;

use crate::application::review_service::{ReviewService, ReviewView, SaveReviewInput};
use crate::error::AppError;

/// Read a media's review. Resolves with the row or null when the user hasn't
/// reviewed it; rejects with an AppError string.
#[command]
pub async fn review_get(
    state: State<'_, SqlitePool>,
    media_id: String,
) -> Result<Option<ReviewView>, AppError> {
    info!(media_id, "review_get invoked");
    let service = ReviewService::new(state.inner().clone());
    service.get(&media_id).await
}

/// Save (create or update) a media's review. Resolves with the stored row (an
/// entirely empty review clears the row and resolves with an empty view) or
/// rejects with an AppError string.
#[allow(clippy::too_many_arguments)]
#[command]
pub async fn review_save(
    state: State<'_, SqlitePool>,
    media_id: String,
    rating: Option<i64>,
    review: Option<String>,
    short_review: Option<String>,
    notes: Option<String>,
    favorite: bool,
    is_spoiler: bool,
    moods: Vec<String>,
    pace: Option<String>,
    content_warnings: Vec<String>,
) -> Result<ReviewView, AppError> {
    info!(media_id, "review_save invoked");
    let service = ReviewService::new(state.inner().clone());
    let input = SaveReviewInput {
        media_id,
        rating,
        review,
        short_review,
        notes,
        favorite,
        is_spoiler,
        moods,
        pace,
        content_warnings,
    };
    service.save(input).await
}

/// Acknowledge a media's current content-warning set (MISSION-079) — stamps
/// `warnings_acknowledged_at` now and resolves with the updated row, or
/// rejects when there is no review / no warnings to acknowledge.
#[command]
pub async fn review_acknowledge_warnings(
    state: State<'_, SqlitePool>,
    media_id: String,
) -> Result<ReviewView, AppError> {
    info!(media_id, "review_acknowledge_warnings invoked");
    let service = ReviewService::new(state.inner().clone());
    service.acknowledge_warnings(&media_id).await
}

/// Delete a media's review row. Resolves or rejects with an AppError string.
#[command]
pub async fn review_delete(state: State<'_, SqlitePool>, media_id: String) -> Result<(), AppError> {
    info!(media_id, "review_delete invoked");
    let service = ReviewService::new(state.inner().clone());
    service.delete(&media_id).await
}
